sdk/storage/Azure.Storage.Blobs/src/PartitionedUploader.cs

```diff
@@ -0,0 +1,449 @@
+                            {
+                                continue;
+                            }
+
+                            await runningTask.ConfigureAwait(false);
+                            runningTasks.RemoveAt(i);
+                            i--;
+                        }
+                    }
+                }
+
+                // Wait for all the remaining blocks to finish staging and then
+                // commit the block list to complete the upload
+                await Task.WhenAll(runningTasks).ConfigureAwait(false);
+                return await _client.CommitBlockListAsync(
+                    blockIds,
+                    blobHttpHeaders,
+                    metadata,
+                    conditions,
+                    accessTier,
+                    cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
+            finally
+            {
+                scope.Dispose();
+            }
+        }
+
+        private async Task StageBlockAsync(
+            StreamPartition block,
+            string blockId,
+            BlobRequestConditions conditions,
+            IProgress<long> progressHandler,
+            CancellationToken cancellationToken)
+        {
+            try
+            {
+                await _client.StageBlockAsync(
+                    blockId,
+                    new MemoryStream(block.Bytes, 0, block.Length, writable: false),
+                    conditions: conditions,
+                    progressHandler: progressHandler,
+                    cancellationToken: cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            finally
+            {
+                // Return the memory used by the block to our ArrayPool as soon
+                // as we've staged it
+                block.Dispose();
+            }
+        }
+
+        // Some streams will throw if you try to access their length so we wrap
+        // the check in a TryGet helper.
+        private static bool TryGetLength(Stream content, out long length)
+        {
+            length = 0;
+            try
+            {
+                if (content.CanSeek)
+                {
+                    length = content.Length;
+                    return true;
+                }
+            }
+            catch (NotSupportedException)
+            {
+            }
+            return false;
+        }
+
+        // Block IDs must be 64 byte Base64 encoded strings
+        private static string GenerateBlockId(long offset)
+        {
+            // TODO #8162 - Add in a random GUID so multiple simultaneous
+            // uploads won't stomp on each other and the first to commit wins.
+            // This will require some changes to our test framework's
+            // RecordedClientRequestIdPolicy.
```

Comment:
I'd rather force some tests to be in live mode, awaiting test framework updates, than have race conditions shipped awaiting that work.

Reply:
That would be almost every blob test.  The old version of parallel upload didn't do this either so I think it's orthogonal to this PR.  Please open a separate work item and have your team decide if this is something you want to push through escrow.  /cc @amishra-dev